File: qt/src/algorithms/OteHuaweiKeygen.cpp
/*
 * Copyright 2013 Rui Araújo, Luís Fonseca
 *
 * This file is part of Router Keygen.
 *
 * Router Keygen is free software: you can redistribute it and/or modify
 * it under the terms of the GNU General Public License as published by
 * the Free Software Foundation, either version 3 of the License, or
 * (at your option) any later version.
 *
 * Router Keygen is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
 * GNU General Public License for more details.
 *
 * You should have received a copy of the GNU General Public License
 * along with Router Keygen.  If not, see <http://www.gnu.org/licenses/>.
 */
#include "OteHuaweiKeygen.h"
#include <QStringList>

OteHuaweiKeygen::OteHuaweiKeygen(QString & ssid, QString & mac,
        QString supported) :
        Keygen(ssid, mac), magicValues(supported) {

}

const int OteHuaweiKeygen::MAGIC_NUMBER = 65535;

QVector<QString> & OteHuaweiKeygen::getKeys() {
    QString mac = getMacAddress();
    if (mac.length() != 12) {
        throw ERROR;
    }
    QStringList magic = magicValues.split(" ");
    QString series = mac.left(2) + mac.mid(6, 2);
    int point;
    if (series == "E8FD" )
        point = 0;
    else if (series == "E8F5")
        point = 1;
    else if (series == "E8F6")
        point = 2;
    else
        return results;
    if (point >= magic.length())
        return results;
    QString pass = "000000" + magic.at(point);
    results.append(pass.mid(pass.length() - 8));
    return results;
}

File: android/routerKeygen/src/main/java/org/exobel/routerkeygen/algorithms/OteHuaweiKeygen.java
package org.exobel.routerkeygen.algorithms;

import android.os.Parcel;
import android.os.Parcelable;

import org.exobel.routerkeygen.R;

import java.util.List;

public class OteHuaweiKeygen extends Keygen {
    public final static int MAGIC_NUMBER = 65535;
    public static final Parcelable.Creator<OteHuaweiKeygen> CREATOR = new Parcelable.Creator<OteHuaweiKeygen>() {
        public OteHuaweiKeygen createFromParcel(Parcel in) {
            return new OteHuaweiKeygen(in);
        }

        public OteHuaweiKeygen[] newArray(int size) {
            return new OteHuaweiKeygen[size];
        }
    };
    private final String magicValues;

    public OteHuaweiKeygen(String ssid, String mac,
                           String magicValues) {
        super(ssid, mac);
        this.magicValues = magicValues;
    }

    private OteHuaweiKeygen(Parcel in) {
        super(in);
        magicValues = in.readString();
    }

    @Override
    public List<String> getKeys() {
        final String mac = getMacAddress();
        if (mac.length() != 12) {
            setErrorCode(R.string.msg_errpirelli);
            return null;
        }
        final String[] magic = magicValues.split(" ");
        final String series = mac.substring(0, 2) + mac.substring(6, 8);
        final int point;
        switch (series) {
            case "E8FD":
                point = 0;
                break;
            case "E8F5":
                point = 1;
                break;
            case "E8F6":
                point = 2;
                break;
            default:
                return getResults();
        }
        if (point >= magic.length)
            return getResults();
        final String pass = "000000" + magic[point];
        addPassword(pass.substring(pass.length() - 8));
        return getResults();
    }

    @Override
    public void writeToParcel(Parcel dest, int flags) {
        super.writeToParcel(dest, flags);
        dest.writeString(magicValues);
    }

}
